engine/lib/filestore.php

```diff
@@ -598,6 +578,230 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = fa
+	}
+
+	$filenameonfilestore = $thumbnail->getFilenameOnFilestore();
+	$last_updated = filemtime($filenameonfilestore);
+	$etag = '"' . $last_updated . '"';
+	$response->setPublic()->setEtag($etag);
+	if ($response->isNotModified($request)) {
+		$response->send();
+		exit();
+	}
+
+	$headers = [
+		'Content-Type' => (new Elgg\Filesystem\MimeTypeDetector())->getType($filenameonfilestore),
+	];
+	$response = new Symfony\Component\HttpFoundation\BinaryFileResponse($filenameonfilestore, 200, $headers, false, 'inline');
+	$response->prepare($request);
+
+	$expires_dt = (new DateTime())->setTimestamp(strtotime('+1 day'));
+	$response->setExpires($expires_dt);
+
+	$response->setEtag($etag);
+	
+	$response->send();
+	exit;
+}
+
+/**
+ * Returns thumbnail sizes configuration for a file entity
+ *
+ * @param \ElggFile $file File entity
+ * @return array
+ */
+function elgg_get_thumbnail_sizes(\ElggFile $file) {
+
+	$sizes = [
+		'small' => [
+			'w' => 60,
+			'h' => 60,
+			'square' => true,
+			'metadata_name' => 'thumbnail',
+			'filename_prefix' => 'thumb',
+		],
+		'medium' => [
+			'w' => 153,
+			'h' => 153,
+			'square' => true,
+			'metadata_name' => 'smallthumb',
+			'filename_prefix' => 'smallthumb',
+		],
+		'large' => [
+			'w' => 600,
+			'h' => 600,
+			'square' => false,
+			'metadata_name' => 'largethumb',
+			'filename_prefix' => 'largethumb',
+		],
+	];
+
+	// @todo: add a hook to filter these values
+	return $sizes;
+}
+
+/**
+ * Attempt to create thumbnails for a file
+ * For non image files (or on a failed attempt to resize source image), all previously generated thumbnails will be cleared
+ *
+ * @param \ElggFile $file File entity
+ * @return bool
+ */
+function elgg_create_thumbnails(\ElggFile $file) {
+
+	if (!$file->guid) {
+		return false;
+	}
+
+	$params = [
+		'entity' => $file,
+	];
+
+	$created = elgg_trigger_plugin_hook('create_thumbnails', 'file', $params, false);
+	if ($created) {
+		$file->icontime = time();
+		return true;
+	}
+
+	if ($file->simpletype !== 'image') {
+		return false;
+	}
+	
+	$prefix = 'file/';
+	$filestorename = elgg_substr($file->getFilename(), elgg_strlen($prefix));
+
+	$sizes = elgg_get_thumbnail_sizes($file);
+
+	foreach ($sizes as $size => $opts) {
+		$metadata_name = $opts['metadata_name'];
+		$filename_prefix = $opts['filename_prefix'];
+		$width = $opts['w'];
+		$height = $opts['h'];
+		$square = $opts['square'];
+		if (!isset($file->$metadata_name)) {
+			// @todo: this logic is flawed and needs to be fixed. See #9612
+			$file->$metadata_name = $prefix . $filename_prefix . $filestorename;
+		}
+		$thumb = elgg_get_thumbnail($file, $size);
+		$thumbnail = get_resized_image_from_existing_file($file->getFilenameOnFilestore(), $width, $height, $square);
+		if ($thumbnail) {
+			$thumb->open("write");
+			$thumb->write($thumbnail);
+			$thumb->close();
+			unset($thumbnail);
+		} else {
+			elgg_clear_thumbnails($file);
+			return false;
+		}
+	}
+
+	$file->icontime = time();
```

Comment:
I am tempted to use filemtime here. It would allow us to eventually move thumb creation to an event handler (so plugins won't have to worry about missing thumbs for file entities generated outside of the upload action). Modified time would serve as a benchmark indicating whether we need to recreate thumbs.